addons/hr_expense/models/hr_expense_sheet.py

```diff
@@ -442,6 +443,20 @@ def _check_expense_lines_company(self):
             if sheet.expense_line_ids.company_id - sheet.company_id:
                 raise ValidationError(_('An expense report must contain only lines from the same company.'))
 
+    @api.onchange('expense_line_ids')
+    def _update_sheet_name(self):
+        expense_lines = self.expense_line_ids
+        if not self.name and expense_lines:
+            if len(expense_lines) == 1:
+                self.name = expense_lines[0].name
+            else:
+                first_date = min(expense.date for expense in expense_lines)
+                last_date = max(expense.date for expense in expense_lines)
+                if first_date == last_date:
+                    self.name = format_date(self.env, first_date)
+                else:
+                    self.name = f"{format_date(self.env, first_date)} - {format_date(self.env, last_date)}"
```

Comment:
duplicate of `_get_default_expense_sheet_values` that would be nice to extract. Especially since your version isn't translatable

```python
    def _get_default_sheet_name(self, expenses_to_report):
        """ Computes the default name for a new expense sheet from the expenses dates """

        dates = expenses_to_report.mapped('date')        
        if False in dates:  # If at least one date isn't set, we don't set a default name
            return False
        
        min_date = format_date(self.env, min(dates))
        max_date = format_date(self.env, max(dates))
        if min_date == max_date:
            sheet_name = min_date
        else:
            sheet_name = _("%(date_from)s - %(date_to)s", date_from=min_date, date_to=max_date)
        return sheet_name
```

and in the _get_default replace it with something like
```python
    def _get_default_expense_sheet_values(self):
        ...
        for todo in sheets:
            paid_by = 'company' if todo[0].payment_mode == 'company_account' else 'employee'
            if len(sheets) > 1:
                sheet_name = _("New Expense Report, paid by %(paid_by)s", paid_by=paid_by)
            else:
                sheet_name = self._get_default_sheet_name(todo)
```

You could even add the "no expenses in self" logic directly inside the new method